Hash token values for storage

jkakavas opened the pull request:
This commit changes how access tokens and refresh tokens are stored
in the tokens index and is a followup to #39631

Access token values are now hashed before being stored in the id
field of the `user_token` and before becoming part of the token
document id. Refresh token values are hashed before being stored
in the token field of the `refresh_token`. The tokens are hashed
without a salt value since these are v4 UUID values that have
enough entropy themselves. Both rainbow table attacks and offline
brute-force attacks are impractical.

As a side effect of this change and in order to support multiple
concurrent refreshes as introduced in #39631, upon refreshing an
<access token, refresh token> pair, the superseding access token
and refresh tokens values are stored in the superseded token doc,
encrypted with a key that is derived from the superseded refresh
token. As such, subsequent requests to refresh the same token in
the predefined time window will return the same superseding access
token and refresh token values, without hitting the tokens index
(as this only stores hashes of the token values). AES in GCM
mode is used for encrypting the token values and the key
derivation from the superseded refresh token uses a small number
of iterations as it needs to be quick.

For backwards compatibility reasons, the new behavior is only
enabled when all nodes in a cluster are in the required version
so that old nodes can cope with the token values in a mixed
cluster during a rolling upgrade.

Resolves #40765

jkakavas committed:
Hash token values for storage

This commit changes how access tokens and refresh tokens are stored
in the tokens index.

Access token values are now hashed before being stored in the id
field of the `user_token` and before becoming part of the token
document id. Refresh token values are hashed before being stored
in the token field of the `refresh_token`. The tokens are hashed
without a salt value since these are v4 UUID values that have
enough entropy themselves. Both rainbow table attacks and offline
bruteforce attacks are impractical.

As a side effect of this change and in order to support multiple
concurrent refreshes as introduced in #39631, upon refreshing an
<access token, refresh token> pair, the superseding access token
and refresh tokens values are stored in the superseded token doc,
encrypted with a key that is derived from the superseded refresh
token. As such, subsequent requests to refresh the same token in
the predefined time window will return the same superseding access
token and refresh token values, without hitting the tokens index
(as this only stores hashes of the token values). AES in GCM
mode is used for encrypting the token values and the key
derivation from the superseded refresh token uses a small number
of iterations as it needs to be quick.

For backwards compatibility reasons, the new behavior is only
enabled when all nodes in a cluster are in the required version
so that old nodes can cope with the token values in a mixed
cluster during a rolling upgrade.

jkakavas committed:
Merge remote-tracking branch 'origin/master' into hash-tokens-storage

jkakavas committed:
fix bug and tests

elasticmachine commented:
Pinging @elastic/es-security

jkakavas commented on x-pack/plugin/core/src/main/resources/security-index-template-7.json:
```diff
@@ -213,8 +213,17 @@
             },
-            "superseded_by": {
-              "type": "keyword"
+            "superseding_access_token": {
```
I'm up for suggestion. I'm not particularly fond of this. It doesn't need to be binary by the way, we can make it text but not index it I guess,

albertzaharovits replied:
I don't see any problem with keeping binary data in a binary type, and I agree that we should set `indexed: false`.

jkakavas replied:
> I don't see any problem with keeping binary data in a binary type, and I agree that we should set indexed: false.

I believe binary fields are not indexed/stored [by default](https://www.elastic.co/guide/en/elasticsearch/reference/current/binary.html) : 

> I don't mind the binary, but I would prefer a nested object rather than having 4 fields 

Keeping binary and will replace this with 
```
"superseding": {
              "type": "object",
              "properties": {
                "access_token": {
                  "type": "binary"
                },
                "refresh_token": {
                  "type": "binary"
                },
                "encryption_iv": {
                  "type": "binary"
                },
                "encryption_salt": {
                  "type": "binary"
                }
              }
            },
```

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -157,11 +158,12 @@
      */
-    private static final int ITERATIONS = 100000;
+    static final int TOKEN_SERVICE_KEY_ITERATIONS = 100000;
+    static final int TOKENS_ENCRYPTION_KEY_ITERATIONS = 1024;
```
The key is derived from a random sting, we don't need too many iterations, we want this to be quick and not computationally expensive.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -235,31 +240,47 @@ public TokenService(Settings settings, Clock clock, Client client, XPackLicenseS
     public void createOAuth2Tokens(Authentication authentication, Authentication originatingClientAuth,
                                    Map<String, Object> metadata, boolean includeRefreshToken,
-                                   ActionListener<Tuple<UserToken, String>> listener) {
+                                   ActionListener<Tuple<String, String>> listener) {
```
This was not required in the context of this PR, but it bothered me for a while because we would call the listener with the UserToken object, when all we need is the access token to return to the user.

albertzaharovits replied:
Thanks for tackling this.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -406,13 +446,19 @@ void decodeToken(String token, ActionListener<UserToken> listener) {
                 // The token was created in a < VERSION_ACCESS_TOKENS_UUIDS cluster so we need to decrypt it to get the tokenId
-                if (in.available() < MINIMUM_BASE64_BYTES) {
-                    logger.debug("invalid token, smaller than [{}] bytes", MINIMUM_BASE64_BYTES);
+                if (in.available() < LEGACY_MINIMUM_BYTES) {
```
This is base64 decoded already so we don't need to check `MINIMUM_BASE64_BYTES`

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
+        // In AES GCM we cannot reuse the same IV. We predictably generate the IV for the second decryption instead of
+        // storing an extra field, since it doesn't have to be unpredictable, just not reused with the same key.
+        byte[] iv2 = new byte[iv.length];
+        System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
```
This has no particular meaning, we just need a predictable way to get a new IV for the second encryption,

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
+    /*
+     * Encrypts the values of the superseding access token and the refresh token, using a key derived from the superseded refresh token.
+     */
+    Tuple<String, String> encryptSupersedingTokens(String supersedingAccessToken, String supersedingRefreshToken,
```
Discarded the idea of using one time pads for encryption as originally discussed because of requirements around FIPS 140-2 compliance

albertzaharovits replied:
Good call!

jkakavas committed:
adjust a couple more tests

albertzaharovits reviewed (commented):
Overall it looks like a simplification, axing the `superseded_by` reference, which is very welcome!

I see that in addition to the hashed token we have to store it encrypted format so that we can return the plaintext on concurrent refreshes. I think that's alright as well.

I would like to propose another simplification, that I believe will lessen the implementation complexity, even though it will require another iteration. I am thinking of completely removing the `refresh_token` and use the `access_token` as a refresh token. That is, removing the `refresh_token` in the token document, the response will include a `refresh_token` field, but getting the token doc from the `refresh_token` would use the same _hash_ and _get_by_id_ . **I think this way we avoid storing two pairs of hashes and encrypted formats, and use only one.**

What do you think? Do you think it would help make this PR leaner?

albertzaharovits commented on x-pack/plugin/core/src/main/resources/security-index-template-7.json:
```diff
@@ -213,8 +213,17 @@
+              "type": "binary"
+            },
+            "superseding_encryption_salt": {
+              "type": "binary"
```
I would move `superseding_encryption_iv` and `superseding_encryption_salt` as data in the refresh token, the same way the standalone `access_tokens` were encrypted before.
The idea is to minimize the number of mapped fields, and also "hide" the implementation details.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -710,44 +756,53 @@ private void indexInvalidation(Collection<String> tokenIds, SecurityIndexManager
-                if (false == refreshTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)
-                        || unencodedRefreshToken.length() != TOKEN_ID_LENGTH) {
-                    logger.debug("Decoded refresh token [{}] with version [{}] is invalid.", unencodedRefreshToken, refreshTokenVersion);
+            if (refreshToken.length() == HASHED_TOKEN_LENGTH) {
```
When is this if branch considered?

jkakavas replied:
This is considered during IDP initiated SAML logout when we need to find relevant tokens based on the token metadata ( that contain the SAML NameID ) and then invalidate those explicitly. See https://github.com/elastic/elasticsearch/blob/9beb31fd3c5a8323cb08cc524f1a2268e9c72c24/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java

jkakavas commented:
Thanks for the :eyes: @albertzaharovits 

>  I would like to propose another simplification, that I believe will lessen the implementation complexity, even though it will require another iteration. I am thinking of completely removing the refresh_token and use the access_token as a refresh token. That is, removing the refresh_token in the token document, the response will include a refresh_token field, but getting the token doc from the refresh_token would use the same hash and get_by_id . I think this way we avoid storing two pairs of hashes and encrypted formats, and use only one.

Could you give it another try to explain the above for my benefit ? I'm not sure I can follow this

jkakavas commented:
> Thanks for the eyes @albertzaharovits
> 
> > I would like to propose another simplification, that I believe will lessen the implementation complexity, even though it will require another iteration. I am thinking of completely removing the refresh_token and use the access_token as a refresh token. That is, removing the refresh_token in the token document, the response will include a refresh_token field, but getting the token doc from the refresh_token would use the same hash and get_by_id . I think this way we avoid storing two pairs of hashes and encrypted formats, and use only one.
> 
> Could you give it another try to explain the above for my benefit ? I'm not sure I can follow this

We discussed this in person with @albertzaharovits and concluded that

- There is no benefit in removing the refresh token and it will potentially differentiate us too much from oAuth2 spec 
- A `superseding` object (not indexed) will be introduced in the mapping to hold all extra fields
- Will attempt to minimize the number of newly introduced fields ( i.e. concatenate `superseding_access_token` and `superseding_refresh_token` to one field before encryption and split values after decryption )

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
+            byte[] iv2 = new byte[iv.length];
+            System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
+            System.arraycopy(iv, 0, iv2, iv.length / 2, iv.length / 2);
+            cipher = getDecryptionCipher(iv2, refreshToken, salt);
```
Per your most recent comment on the PR, I think it definitely makes sense to do a single encrypt of the concatenated tokens.

tvernum reviewed (commented):
I haven't reviewed the tests yet, but since you're planning a couple of changes I'll submit my main comments now.

I'm starting to get concerned about the number of places we're passing around `Tuple<String,String>` and it doesn't always mean the same thing.
Sometimes the 2nd string is a plain refresh token, sometimes it's a hashed refresh token. Maybe there's other combinations, it's hard to tell.
I don't have a preferred solution, but I think types are important, and the code is getting harder to follow because we're not clearly defining those types.

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -283,15 +313,16 @@ private void createOAuth2Tokens(String userTokenId, Version tokenVersion, Securi
                                         // prior versions are not version-prepended, as nodes on those versions don't expect it.
                                         // Such nodes might exist in a mixed cluster during a rolling upgrade.
-                                        listener.onResponse(new Tuple<>(userToken, plainRefreshToken));
+                                        listener.onResponse(new Tuple<>(versionedAccessToken, plainRefreshToken));
```
Yes, this is correct. The comment refers to the refresh token (.. too many versions and possibilities to keep track of ). I've updated the comment

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -235,31 +240,47 @@ public TokenService(Settings settings, Clock clock, Client client, XPackLicenseS
+     * tokens are stored in the security index for versions up to 7.1.0 and to a specific security tokens index for later versions.
      */
-    private void createOAuth2Tokens(String userTokenId, Version tokenVersion, SecurityIndexManager tokensIndex,
+    private void createOAuth2Tokens(String accessToken, String plainRefreshToken, Version tokenVersion, SecurityIndexManager tokensIndex,
```
I think the main reason is that the refresh tokens can be non versioned (pre v7.1.0), but the access tokens are already versioned. Maybe it's time for `AccessToken` and `RefreshToken` too

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectAuthenticateAction.java:
```diff
@@ -36,6 +38,7 @@
     private final ThreadPool threadPool;
     private final AuthenticationService authenticationService;
     private final TokenService tokenService;
+    private static final Logger logger = LogManager.getLogger(TransportOpenIdConnectLogoutAction.class);
```
Used in line 78 (but the `getLogger()` argument was wrong :/ ) . There is also a `logger` in  `TransportAction` but it's use is deprecated

jkakavas committed:
address feedback

jkakavas committed:
Address feedback

- Move superseding encrypted token data in a separate object
- Concatenate tokens before encryption and split after decryption
  so that we only have 1 crypto operation and one less field in
  the mapping
- Add javadoc

jkakavas committed:
fix BWC tests

jkakavas committed:
Merge remote-tracking branch 'origin/master' into hash-tokens-storage

jkakavas committed:
re-enable handle of hashed refresh tokens when searching for token documents

jkakavas committed:
Checkstyle violations

jkakavas commented:
> I'm starting to get concerned about the number of places we're passing around Tuple<String,String> and it doesn't always mean the same thing.

All instances used are Tuples containing the serialized access token and the serialized refresh token, as these will be returned to the caller of our APIs. I've updated the javadoc where applicable to denote this.

jkakavas commented:
@tvernum this is ready for a review round

jkakavas committed:
Merge remote-tracking branch 'origin/master' into hash-tokens-storage

tvernum commented:
> All instances used are Tuples containing the serialized access token and the serialized refresh token, 

In line 1420 we return a hashedRefreshToken, but I _think_ the other cases are all unhashed (but versioned, unless in BWC mode).

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -1595,21 +1673,31 @@ private Cipher getDecryptionCipher(byte[] iv, SecretKey key, Version version, By
+    }
+
+    byte[] getRandomBytes(int length) {
+        final byte[] initializationVector = new byte[length];
```
Nit
```suggestion
        final byte[] bytes = new byte[length];
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -214,16 +209,21 @@ public void testInvalidAuthorizationHeader() throws Exception {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
+        // This test only makes sense in mixed clusters with pre v7.1.0 nodes where the Key is actually used
+        if (null == oldNode) {
+            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_0_0));
```
Minor, this feature is now in v7.2, not v7.1
```suggestion
            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_1_0));
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -267,27 +270,32 @@ private void rotateKeys(TokenService tokenService) {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
+        // This test only makes sense in mixed clusters with pre v7.1.0 nodes where the Key is actually used
+        if (null == oldNode) {
+            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_0_0));
```
```suggestion
            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_1_0));
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -298,22 +306,27 @@ public void testKeyExchange() throws Exception {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
+        // This test only makes sense in mixed clusters with pre v7.1.0 nodes where the Key is actually used
+        if (null == oldNode) {
+            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_0_0));
```
```suggestion
            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_1_0));
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -366,16 +382,21 @@ public void testPruneKeys() throws Exception {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
+        // This test only makes sense in mixed clusters with pre v7.1.0 nodes where the Key is actually used
+        if (null == oldNode) {
+            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_0_0));
```
```suggestion
            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_1_0));
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -395,29 +416,40 @@ public void testPassphraseWorks() throws Exception {
         TokenService tokenService = createTokenService(tokenServiceEnabledSettings, systemUTC());
+        // This test only makes sense in mixed clusters with pre v7.1.0 nodes where the Key is actually used
+        if (null == oldNode) {
+            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_0_0));
```
```suggestion
            oldNode = addAnotherDataNodeWithVersion(this.clusterService, randomFrom(Version.V_6_7_0, Version.V_7_1_0));
```

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenServiceTests.java:
```diff
@@ -620,34 +656,64 @@ public void testIndexNotAvailable() throws Exception {
+        assertAuthentication(authentication, retrievedAuth);
+    }
+
+    public void testTokenEncryption() throws Exception {
```
```suggestion
    public void testSupercedingTokenEncryption() throws Exception {
```

tvernum reviewed (approved):
LGTM, with a few minor nits.

jkakavas committed:
address review
